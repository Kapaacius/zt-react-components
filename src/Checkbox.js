import React, {PureComponent} from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import {borderRadiuses, colors, fontSizes} from "./designGuideline";
import Checkmark from "./icons/checkmark.svg";

const CheckboxElement = styled.span`
    display: inline-flex;
    align-items: center;
    justify-content: center;

    margin-right: 7px;

    border: 1px solid ${({checked}) => checked ? colors.black : colors.gray};
    border-radius: ${borderRadiuses.default};

    background-color: ${({disabled}) => disabled ? colors.darkerWhite : colors.white};

    width: ${({size}) => size === "large" ? fontSizes.slightlyLarger : fontSizes.default};
    height: ${({size}) => size === "large" ? fontSizes.slightlyLarger : fontSizes.default};
    font-size: ${({size}) => size === "large" ? fontSizes.slightlyLarger : fontSizes.default};

    label:hover > &:not([disabled]) {
        border-color: #000;
    }
`;

export default class Checkbox extends PureComponent {
    static propTypes = {
        name: PropTypes.string,
        // small or large
        size: PropTypes.string,
        checked: PropTypes.bool,
        disabled: PropTypes.bool
    };

    render() {
        return (
            <CheckboxElement
                checked={this.props.checked}
                size={this.props.size}
                disabled={this.props.disabled}
            >
                {this.props.checked && <Checkmark />}
                {this.props.name && this.renderHiddenInput()}
            </CheckboxElement>
        );
    }

    renderHiddenInput() {
        return (
            <input
                type="hidden"
                name={this.props.name}
                value={this.props.checked}
                onClick={(event) => event.stopPropagation()}
            />
        );
    }
}
