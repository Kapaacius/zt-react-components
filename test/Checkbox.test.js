import "jest-styled-components";
import React from "react";
import renderer from "react-test-renderer";
import Checkbox from "../src/Checkbox";

describe("<Checkbox />", function() {
    it("renders styled span element with hidden input", function() {
        expect(renderer.create(<Checkbox name="test" />)).toMatchSnapshot();
    });

    it("renders checked", function() {
        expect(renderer.create(<Checkbox checked={true} />)).toMatchSnapshot();
    });

    it("renders disabled", function() {
        expect(renderer.create(<Checkbox disabled={true} />)).toMatchSnapshot();
    });

    it("renders larger", function() {
        expect(renderer.create(<Checkbox size="large" />)).toMatchSnapshot();
    });

    it("stops hidden input event propagation on click", function() {
        const checkbox = renderer.create(<Checkbox name="test" />);
        const stopPropagation = jasmine.createSpy("stopPropagation");

        checkbox.toJSON().children[0].props.onClick({stopPropagation});

        expect(stopPropagation).toHaveBeenCalled();
    });
});
