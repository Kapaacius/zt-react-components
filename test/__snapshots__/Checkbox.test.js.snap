// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> renders checked 1`] = `
.gLVrHx {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin-right: 7px;
  border: 1px solid #000;
  border-radius: 4px;
  background-color: #fff;
  width: 12px;
  height: 12px;
  font-size: 12px;
}

<span
  checked={true}
  className="sc-bdVaJa gLVrHx"
  disabled={undefined}
  size={undefined}
>
  <svg
    height="7"
    viewBox="0 0 9 7"
    width="9"
    xmlns="http://www.w3.org/2000/svg"
  >
    <title>
      P
    </title>
    <path
      d="M8.807.477A.625.625 0 0 0 8.35.283a.57.57 0 0 0-.44.194L3.217 5.17 1.09 3.043a.596.596 0 0 0-.44-.176.657.657 0 0 0-.457.176A.625.625 0 0 0 0 3.5c0 .176.064.328.193.457L2.76 6.523a.625.625 0 0 0 .457.194.57.57 0 0 0 .44-.194l5.15-5.15a.615.615 0 0 0 0-.896z"
      fill="#2F333D"
      fillRule="evenodd"
    />
  </svg>
</span>
`;

exports[`<Checkbox /> renders disabled 1`] = `
.hTZnim {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin-right: 7px;
  border: 1px solid #999;
  border-radius: 4px;
  background-color: #f3f3f3;
  width: 12px;
  height: 12px;
  font-size: 12px;
}

<span
  checked={undefined}
  className="sc-bdVaJa hTZnim"
  disabled={true}
  size={undefined}
/>
`;

exports[`<Checkbox /> renders larger 1`] = `
.hnoCGB {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin-right: 7px;
  border: 1px solid #999;
  border-radius: 4px;
  background-color: #fff;
  width: 14px;
  height: 14px;
  font-size: 14px;
}

<span
  checked={undefined}
  className="sc-bdVaJa hnoCGB"
  disabled={undefined}
  size="large"
/>
`;

exports[`<Checkbox /> renders styled span element with hidden input 1`] = `
.ldAYDL {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin-right: 7px;
  border: 1px solid #999;
  border-radius: 4px;
  background-color: #fff;
  width: 12px;
  height: 12px;
  font-size: 12px;
}

<span
  checked={undefined}
  className="sc-bdVaJa ldAYDL"
  disabled={undefined}
  size={undefined}
>
  <input
    name="test"
    onClick={[Function]}
    type="hidden"
    value={undefined}
  />
</span>
`;
